dhall/src/Dhall/Format.hs

```diff
@@ -117,3 +131,100 @@ format (Format { input = input0, ..}) = go input0
                         let modified = "formatted"
 
                         Control.Exception.throwIO CheckFailed{..}
+
+--
+detectCharacterSet :: Expr Src a -> CharacterSet
+detectCharacterSet = go mempty
+  where
+    recurse = go ASCII
+
+    -- Short-circuit traversal if Unicode encountered
+    go Unicode = \_ -> Unicode
+    go ASCII = \case
+        Note Src{srcText} expr ->
+            case expr of
+                Lam _ _
+                    | "λ" `Text.isInfixOf` srcText ||
+                      "→" `Text.isInfixOf` srcText -> Unicode
+                Pi _ _ _
+                    | "∀" `Text.isInfixOf` srcText ||
+                      "→" `Text.isInfixOf` srcText -> Unicode
+                Prefer _ _ _
+                    | "⫽" `Text.isInfixOf` srcText -> Unicode
+                Combine _ _ _
+                    | "∧" `Text.isInfixOf` srcText -> Unicode
+                CombineTypes _ _
+                    | "⩓" `Text.isInfixOf` srcText -> Unicode
+                _ -> recurse expr
+
+        -- Recurse down the AST
```

Comment:
Do we have a better way to recurse on the rest of the AST?

Reply:
We do!  You can use the `subExpressions` traversal in conjunction with one of the utilities from `Control.Lens.Plated`.  For example, I think this would work:

```haskell
detectCharacterSet :: Expr s a -> CharacterSet
detectCharacterSet = foldOf (cosmosOf subExpressions . to characterSet)
  where
    characterSet :: Expr s a -> CharacterSet
    characterSet (Note Src{ srcText } expression) = case expression of
        Lam{} | … -> Unicode
        Pi{} | … -> Unicode
        …
        CombineTypes{} | …-> Unicode
        _ -> mempty
```

That way you don't need to implement the recursion.  The `characterSet` function only needs to analyze things one level deep.

Note that this might require vendoring the implementation of `cosmosOf` in the `Dhall.Optics` module to avoid the `lens` dependency

Reply:
Woah that looks like magic 😍

I'll give it a go! 👍

Reply:
It works very well, thank you! I need to learn these advanced lenses one day :sweat_smile:

Reply:
I think I missed the importance of the guards in your suggestion. My current version probably doesn't handle nested Lambdas and only uses the character set from the outer one.

I will fix that and I will also try to add some tests to cover the new behaviour of dhall format.